File: android/truepad-spt/src/main/kotlin/dev/systemslibrarian/truepad/spt/XWing.kt
package dev.systemslibrarian.truepad.spt

import org.bouncycastle.pqc.crypto.xwing.XWingKEMExtractor
import org.bouncycastle.pqc.crypto.xwing.XWingKEMGenerator
import org.bouncycastle.pqc.crypto.xwing.XWingKeyGenerationParameters
import org.bouncycastle.pqc.crypto.xwing.XWingKeyPairGenerator
import org.bouncycastle.pqc.crypto.xwing.XWingPrivateKeyParameters
import org.bouncycastle.pqc.crypto.xwing.XWingPublicKeyParameters
import java.security.SecureRandom

/* ============================================================================
 * X-Wing (suite 0x0001) — the narrow TruePad wrapper (Kotlin/Android twin)
 * ----------------------------------------------------------------------------
 * A BYTE-EXACT twin of src/spt/xwing-v1.ts. docs/SEALED-PAD-TRANSFER.md §2.2
 * freezes the WHOLE of draft-connolly-cfrg-xwing-kem-10 as suite 0x0001. This
 * file does not implement X-Wing; it wraps Bouncy Castle's low-level
 * `org.bouncycastle.pqc.crypto.xwing` KEM, whose construction the audit
 * (docs/SEALED-PAD-TRANSFER-VALIDATION.md; the Android crypto-library audit)
 * verified byte-identical to draft-10 / @noble/post-quantum 0.7.1 — same seed,
 * SHAKE256(96) expansion, ML-KEM-coins-then-X25519 randomness order, and the
 * SHA3-256(ss_M ‖ ss_X ‖ ct_X ‖ pk_X ‖ label) combiner with the label at the end.
 *
 * It adds NO cryptography (no combiner, no extra KDF), exactly as the TS wrapper
 * refuses to. It uses BC through its LOW-LEVEL API, never registering BC as a JCA
 * provider, so it cannot clash with Android's bundled org.bouncycastle.
 *
 * The 32-byte SEED is the persisted private key (§12), never an expanded,
 * implementation-specific key structure — the reason a recipient key stays
 * portable across any conforming X-Wing implementation.
 *
 * Randomness is injected as a SecureRandom so the KAT corpus can drive the KEM
 * derandomized (draft-10 Appendix C). A PRODUCTION caller MUST use the real
 * platform CSPRNG: passing a chosen RNG to a KEM chooses the shared secret too.
 * ========================================================================= */

const val XWING_SEED_BYTES = 32
const val XWING_PUBLIC_KEY_BYTES = 1216
const val XWING_CIPHERTEXT_BYTES = 1120
const val XWING_SHARED_SECRET_BYTES = 32

/** `eseed` is 64 bytes: [0,32) the ML-KEM-768 coins, [32,64) the X25519 ephemeral
 *  scalar (§2.2). Only meaningful for the derandomized KAT path. */
const val XWING_ESEED_BYTES = 64

class XWingKeyPair(val decapsulationSeed: ByteArray, val encapsulationKey: ByteArray)

class XWingEncapsulation(val ciphertext: ByteArray, val sharedSecret: ByteArray)

private fun requireLength(bytes: ByteArray, expected: Int, what: String) {
    if (bytes.size != expected) throw IllegalArgumentException("$what: expected $expected bytes, got ${bytes.size}")
}

object XWing {

    /** Key generation. Production callers omit [rng] (fresh platform CSPRNG). The
     *  KAT passes a fixed RNG that returns the 32-byte seed so `pk` is reproducible. */
    fun generateKeyPair(rng: SecureRandom = SecureRandom()): XWingKeyPair {
        val gen = XWingKeyPairGenerator()
        gen.init(XWingKeyGenerationParameters(rng))
        val kp = gen.generateKeyPair()
        val seed = (kp.private as XWingPrivateKeyParameters).seed
        val encapsulationKey = (kp.public as XWingPublicKeyParameters).encoded
        requireLength(seed, XWING_SEED_BYTES, "decapsulationSeed")
        requireLength(encapsulationKey, XWING_PUBLIC_KEY_BYTES, "encapsulationKey")
        return XWingKeyPair(seed, encapsulationKey)
    }

    /** Encapsulate to [encapsulationKey]. Production callers omit [rng]. The KAT
     *  passes a fixed RNG returning the 64-byte `eseed` so `ct`/`ss` reproduce. */
    fun encapsulate(encapsulationKey: ByteArray, rng: SecureRandom = SecureRandom()): XWingEncapsulation {
        requireLength(encapsulationKey, XWING_PUBLIC_KEY_BYTES, "encapsulationKey")
        val pk = XWingPublicKeyParameters(encapsulationKey)
        val out = XWingKEMGenerator(rng).generateEncapsulated(pk)
        val ciphertext = out.encapsulation
        val sharedSecret = out.secret
        requireLength(ciphertext, XWING_CIPHERTEXT_BYTES, "ciphertext")
        requireLength(sharedSecret, XWING_SHARED_SECRET_BYTES, "sharedSecret")
        return XWingEncapsulation(ciphertext, sharedSecret)
    }

    /** Decapsulate — fully deterministic given [ciphertext] and the 32-byte seed. */
    fun decapsulate(ciphertext: ByteArray, decapsulationSeed: ByteArray): ByteArray {
        requireLength(ciphertext, XWING_CIPHERTEXT_BYTES, "ciphertext")
        requireLength(decapsulationSeed, XWING_SEED_BYTES, "decapsulationSeed")
        val sk = XWingPrivateKeyParameters(decapsulationSeed)
        val sharedSecret = XWingKEMExtractor(sk).extractSecret(ciphertext)
        requireLength(sharedSecret, XWING_SHARED_SECRET_BYTES, "sharedSecret")
        return sharedSecret
    }
}

File: ios/TruePadKit/Sources/TruePadSPT/XWing.swift
/* ============================================================================
 * X-Wing (suite 0x0001) — the narrow TruePad wrapper
 * ----------------------------------------------------------------------------
 * Byte-exact twin of src/spt/xwing-v1.ts.
 *
 * docs/SEALED-PAD-TRANSFER.md §2.2 freezes the WHOLE of
 * draft-connolly-cfrg-xwing-kem-10 as suite 0x0001. This file does not implement
 * it; it wraps swift-crypto's `XWingMLKEM768X25519`, whose construction is
 * BoringSSL's and is validated here against the draft's own Appendix C vectors
 * (XWingKATTests) and the cross-language SPT interop corpus.
 *
 * What this wrapper is FOR, given the library already exposes a KEM:
 *
 *   · it pins the sizes, so a library change that altered any of them fails here
 *     rather than three layers up;
 *   · it fixes TruePad's persisted private key as the 32-byte X-Wing SEED and
 *     never an expanded, implementation-specific key structure — the reason a
 *     recipient key stays portable across any conforming X-Wing implementation,
 *     and across editions;
 *   · it keeps encapsulation behind a protocol, so tests can supply the
 *     derandomized variant the reference vectors need WITHOUT that capability
 *     existing in a shipping build (see TruePadKATSupport).
 *
 * It adds NO cryptography. There is no TruePad combiner, no extra KDF, and no
 * all-zero X25519 policy of our own: adding any of those would take suite 0x0001
 * outside the construction that was analysed, which §2.2 refuses in terms.
 * BoringSSL already rejects a degenerate X25519 result, matching the Browser and
 * Android Editions; XWingHostileInputTests pins that as a contract.
 * ========================================================================= */

import Crypto
import Foundation

public struct XWingKeyPair: Sendable {
    /// The 32-byte X-Wing seed. THIS is the recipient's private key; everything
    /// else is re-derivable.
    public let decapsulationSeed: [UInt8]
    public let encapsulationKey: [UInt8]
}

public struct XWingEncapsulation: Sendable {
    public let ciphertext: [UInt8]
    public let sharedSecret: [UInt8]

    public init(ciphertext: [UInt8], sharedSecret: [UInt8]) {
        self.ciphertext = ciphertext
        self.sharedSecret = sharedSecret
    }
}

/// How a shared secret is encapsulated to a recipient.
///
/// Production has exactly one conformer, `SystemXWingEncapsulator`, which draws
/// its entropy from the system CSPRNG. The protocol exists so the deterministic
/// variant needed by the frozen reference vectors can be supplied by TESTS
/// without any production type ever accepting an `eseed`: note that the method
/// below takes the recipient's key and nothing else. There is deliberately no
/// entropy parameter anywhere in this module.
public protocol XWingEncapsulating: Sendable {
    func encapsulate(encapsulationKey: [UInt8]) throws -> XWingEncapsulation
}

public enum XWing {
    static func requireLength(_ bytes: [UInt8], _ expected: Int, _ what: String) throws {
        guard bytes.count == expected else {
            throw SptError.wrongLength(what, expected: expected, got: bytes.count)
        }
    }

    /// Production key generation. Randomness comes from the platform CSPRNG;
    /// §13 forbids a seed argument on this path.
    public static func generateKeyPair() throws -> XWingKeyPair {
        let priv = try XWingMLKEM768X25519.PrivateKey.generate()
        let seed = [UInt8](priv.seedRepresentation)
        let pub = [UInt8](priv.publicKey.rawRepresentation)
        try requireLength(seed, SptConstants.xwingSeedBytes, "decapsulationSeed")
        try requireLength(pub, SptConstants.xwingPublicKeyBytes, "encapsulationKey")
        return XWingKeyPair(decapsulationSeed: seed, encapsulationKey: pub)
    }

    /// Re-derive the encapsulation key from a stored seed. The seed is the
    /// persisted private key; this is how a recipient recovers its public half
    /// after a restart without having stored it.
    public static func publicKey(fromSeed seed: [UInt8]) throws -> [UInt8] {
        try requireLength(seed, SptConstants.xwingSeedBytes, "decapsulationSeed")
        let priv = try XWingMLKEM768X25519.PrivateKey(seedRepresentation: seed, publicKey: nil)
        let pub = [UInt8](priv.publicKey.rawRepresentation)
        try requireLength(pub, SptConstants.xwingPublicKeyBytes, "encapsulationKey")
        return pub
    }

    public static func decapsulate(ciphertext: [UInt8], decapsulationSeed: [UInt8]) throws -> [UInt8] {
        try requireLength(ciphertext, SptConstants.xwingCiphertextBytes, "ciphertext")
        try requireLength(decapsulationSeed, SptConstants.xwingSeedBytes, "decapsulationSeed")
        let priv = try XWingMLKEM768X25519.PrivateKey(
            seedRepresentation: decapsulationSeed, publicKey: nil)
        let ss = try priv.decapsulate(Data(ciphertext))
        let bytes = ss.withUnsafeBytes { [UInt8]($0) }
        try requireLength(bytes, SptConstants.xwingSharedSecretBytes, "sharedSecret")
        return bytes
    }
}

/// Production encapsulation. Randomness comes from the platform CSPRNG inside
/// swift-crypto; §13 forbids a seed argument on this path.
public struct SystemXWingEncapsulator: XWingEncapsulating {
    public init() {}

    public func encapsulate(encapsulationKey: [UInt8]) throws -> XWingEncapsulation {
        try XWing.requireLength(encapsulationKey,
                                SptConstants.xwingPublicKeyBytes, "encapsulationKey")
        let pub = try XWingMLKEM768X25519.PublicKey(rawRepresentation: encapsulationKey)
        let result = try pub.encapsulate()
        let ct = [UInt8](result.encapsulated)
        let ss = result.sharedSecret.withUnsafeBytes { [UInt8]($0) }
        try XWing.requireLength(ct, SptConstants.xwingCiphertextBytes, "ciphertext")
        try XWing.requireLength(ss, SptConstants.xwingSharedSecretBytes, "sharedSecret")
        return XWingEncapsulation(ciphertext: ct, sharedSecret: ss)
    }
}
